src/Behaviors/Meshes/baseSixDofDragBehavior.ts

```diff
@@ -0,0 +1,378 @@
+                            this._attachedToElement = true;
+                        } else {
+                            this._attachedToElement = false;
+                        }
+                    }
+
+                    this._targetDragStart(virtualMeshesInfo.dragMesh.position, virtualMeshesInfo.dragMesh.rotationQuaternion!, pointerId);
+                    this.onDragStartObservable.notifyObservers({ position: virtualMeshesInfo.dragMesh.position });
+                }
+            } else if (pointerInfo.type == PointerEventTypes.POINTERUP || pointerInfo.type == PointerEventTypes.POINTERDOUBLETAP) {
+                const registeredPointerIndex = this.currentDraggingPointerIds.indexOf(pointerId);
+
+                // Update state
+                virtualMeshesInfo.dragging = false;
+
+                if (registeredPointerIndex !== -1) {
+                    this.currentDraggingPointerIds.splice(registeredPointerIndex, 1);
+                    if (this.currentDraggingPointerIds.length === 0) {
+                        this._moving = false;
+                        this._dragging = false;
+
+                        // Reattach camera controls
+                        if (this.detachCameraControls && this._attachedToElement && this._pointerCamera && !this._pointerCamera.leftCamera) {
+                            this._pointerCamera.attachControl(true);
+                            this._attachedToElement = false;
+                        }
+                    }
+
+                    virtualMeshesInfo.originMesh.removeChild(virtualMeshesInfo.dragMesh);
+                    virtualMeshesInfo.originMesh.removeChild(virtualMeshesInfo.pivotMesh);
+                    this._targetDragEnd(pointerId);
+                    this.onDragEndObservable.notifyObservers({});
+                }
+            } else if (pointerInfo.type == PointerEventTypes.POINTERMOVE) {
+                const registeredPointerIndex = this.currentDraggingPointerIds.indexOf(pointerId);
+
+                if (registeredPointerIndex !== -1 && virtualMeshesInfo.dragging && pointerInfo.pickInfo && pointerInfo.pickInfo.ray) {
+                    let zDragFactor = this.zDragFactor;
+
+                    // 2 pointer interaction should not have a z axis drag factor
+                    if (this.currentDraggingPointerIds.length > 1) {
+                        zDragFactor = 0;
+                    }
+
+                    if (this._pointerCamera && this._pointerCamera.cameraRigMode == Camera.RIG_MODE_NONE && !this._pointerCamera._isLeftCamera && !this._pointerCamera._isRightCamera) {
+                        pointerInfo.pickInfo.ray.origin.copyFrom(this._pointerCamera!.globalPosition);
+                        zDragFactor = 0;
+                    }
+
+                    // Calculate controller drag distance in controller space
+                    var originDragDifference = pointerInfo.pickInfo.ray.origin.subtract(virtualMeshesInfo.lastOriginPosition);

```

Comment:
could you use subtractToRef with temp vectors?

If not - (nit) - change to `const`